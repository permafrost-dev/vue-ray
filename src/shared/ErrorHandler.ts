/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */

import VueRay, { ray } from '../VueRay';
import { format as prettyFormat } from '@permafrost-dev/pretty-format';

export type ERROR_TYPE = 'vue' | 'error' | 'unhandled_rejection';

export type AdditionalErrorInfoCallback = (err: any | null, ctx: any | null, type: ERROR_TYPE) => any;

export class ErrorHandler {
    public window: any;
    public ray: VueRay | any;

    public additionalInfoCallback: AdditionalErrorInfoCallback | null = null;

    constructor(callback: AdditionalErrorInfoCallback | null = null, window: any = null, rayInstance: any = null) {
        this.additionalInfoCallback = callback;
        this.ray = rayInstance ?? ray();
        this.window = window ?? globalThis;
    }

    public additionalInfo(err: any, ctx: any, type: ERROR_TYPE): any {
        return this.additionalInfoCallback ? this.additionalInfoCallback(err, ctx, type) : '';
    }

    public additionalInfoHtml(err: any, ctx: any, type: ERROR_TYPE): string {
        let info = this.additionalInfo(err, ctx, type);

        if (typeof info === 'string' && !info.length) {
            return '';
        }

        if (typeof info !== 'string' && typeof info !== 'number') {
            info = prettyFormat(info, { indent: 4, highlight: true });

            return (
                `<details open><summary class="text-black w-full block mt-1 cursor-pointer">Additional Info:</summary>` +
                `<div class="text-blue-600 w-full block pl-8">${info}</div></details>`
            );
        }

        return (
            `<div class="text-black w-full block mt-1">Additional Info:</div>` +
            `<div class="text-blue-600 w-full block pl-8">${info}</div>`
        );
    }

    public onVueError = (err: any, vm: any) => {
        const additionalInfoHtml = this.additionalInfoHtml(err, vm, 'vue');

        // pretty-print the error message and stack trace, and append any additional info provided by the callback
        const stack = err.stack
            .replace(
                /^(\w+):(.+)$/m,
                '<div class="flex bg-white border-0">' +
                    '<div class="flex-row bg-white text-red-400 pr-1 pl-1 border border-1 border-red-400">$1</div>' +
                    '<div style="padding-top: 0.25em;" class="flex-row">$2</div>' +
                    '</div>',
            )
            .replace(/\n/, 'Stack Trace:\n')
            .replace(/ at ([^ ]+) (.*)$/gm, '<div class="inline"> at <span class="text-blue-700">$1</span> $2</div>'); // eslint-disable-line

        vm.$ray().sendCustom(`<pre>${stack}${additionalInfoHtml}</pre>`, 'Error')
            .small()
            .red();
    };

    public onWindowError = (message: string, source: string, lineno: number, colno: number, error: any) => {
        const additionalInfoHtml = this.additionalInfoHtml(error, { message, source, lineno, colno }, 'error');

        const html = `
        <div class="flex bg-white border-0">
            <div class="flex-row bg-white text-red-400 pr-1 pl-1 border border-1 border-red-400">${message}</div>
            <div style="padding-top: 0.25em;" class="flex-row"><code>${source}</code></div>
        </div>${additionalInfoHtml}`;

        this.ray?.html(`<pre>${html}</pre>`, 'Error').small()
            .red();
    };

    public onWindowUnhandledRejectionEvent = (event: any) => {
        const additionalInfoHtml = this.additionalInfoHtml(null, event, 'unhandled_rejection');

        const html = `
        <div class="flex bg-white border-0">
            <div class="flex-row bg-white text-red-400 pr-1 pl-1 border border-1 border-red-400">${event.reason}</div>
            <div style="padding-top: 0.25em;" class="flex-row"><code>${JSON.stringify(event)}</code></div>
        </div>${additionalInfoHtml}`;

        this.ray?.html(`<pre>${html}</pre>`, 'Error').small()
            .red();
    };

    public installVueErrorHandler(vueConfig: any): any {
        vueConfig.errorHandler = this.onVueError;

        return vueConfig;
    }

    public installWindowErrorHandlers(onError: any = null, onUnhandled: any = null) {
        if (this.window) {
            this.window.onerror = onError ?? this.onWindowError;
            this.window.unhandledrejection = onUnhandled ?? this.onWindowUnhandledRejectionEvent;
        }
    }
}

export const errorHandler = new ErrorHandler();

export default errorHandler;
