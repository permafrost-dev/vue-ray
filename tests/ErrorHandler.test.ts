/* eslint-disable no-unused-vars */
import { ErrorHandler } from '@/ErrorHandler';
import { beforeEach, describe, expect, it } from 'vitest';
import { FakeRay } from './TestClasses/FakeRay';

let options: any, window: any, rayInstance: FakeRay, vueConfig: any, errorHandler: ErrorHandler;

beforeEach(() => {
    options = {};
    vueConfig = { errorHandler: null };
    window = { onerror: null, unhandledrejection: null };
    rayInstance = new FakeRay();
});

describe('Error Handler:', () => {
    it('returns additional info from the callback', () => {
        errorHandler = new ErrorHandler((...args) => 'test', window, rayInstance);

        expect(errorHandler.additionalInfo(null, null, 'error')).toBe('test');
    });

    it('returns an empty string for additional info when the callback is null', () => {
        errorHandler = new ErrorHandler(null, window, rayInstance);

        expect(errorHandler.additionalInfo(null, null, 'error')).toBe('');
    });

    it('returns additional info html', () => {
        errorHandler = new ErrorHandler((...args) => 'test', window, rayInstance);

        const html = errorHandler.additionalInfoHtml(null, null, 'error');

        expect(html).toContain('<div');
        expect(html).toContain('</div>');
        expect(html).toContain('>test<');
    });

    it('returns an empty string as additional info html when the callback is null', () => {
        errorHandler = new ErrorHandler(null, window, rayInstance);

        expect(errorHandler.additionalInfoHtml(null, null, 'error')).toBe('');
    });

    it('sends an html payload and appends the callback result when handling vue errors', () => {
        errorHandler = new ErrorHandler((...args) => '---CUSTOM_DATA---', window, rayInstance);

        errorHandler.onVueError(
            { stack: '' },
            {
                $ray(...args: any[]) {
                    return rayInstance.send(...args);
                },
            },
        );

        expect(rayInstance.payloads).toMatchSnapshot();
        expect(rayInstance.payloads[0].content).toContain('---CUSTOM_DATA---');
    });

    it('sends an html payload and appends the callback result when handling window errors', () => {
        errorHandler = new ErrorHandler((...args) => '---CUSTOM_DATA---', window, rayInstance);

        errorHandler.onWindowError('message', 'source', 1, 1, {});

        expect(rayInstance.payloads).toMatchSnapshot();
        expect(rayInstance.payloads[0].content).toContain('---CUSTOM_DATA---');
    });

    it('sends an html payload and appends the callback result when handling window unhandled rejection event errors', () => {
        errorHandler = new ErrorHandler((...args) => '---CUSTOM_DATA---', window, rayInstance);

        errorHandler.onWindowUnhandledRejectionEvent({ reason: 'some-data' });

        expect(rayInstance.payloads).toMatchSnapshot();
        expect(rayInstance.payloads[0].content).toContain('---CUSTOM_DATA---');
    });

    it('installs the vue error handler', () => {
        errorHandler = new ErrorHandler(null, window, rayInstance);

        expect(vueConfig.errorHandler).toBeNull();

        errorHandler.installVueErrorHandler(vueConfig);

        expect(vueConfig.errorHandler).not.toBeNull();
    });

    it('installs the default window error handlers', () => {
        errorHandler = new ErrorHandler(null, window, rayInstance);

        errorHandler.installWindowErrorHandlers(null, null);

        expect(window.onerror).toBe(errorHandler.onWindowError);
        expect(window.unhandledrejection).toBe(errorHandler.onWindowUnhandledRejectionEvent);
    });

    it('installs custom window error handlers', () => {
        errorHandler = new ErrorHandler(null, window, rayInstance);

        const customErrorHandler1 = function (...args: any[]) {
            return false;
        };

        const customErrorHandler2 = function (...args: any[]) {
            return false;
        };

        errorHandler.installWindowErrorHandlers(customErrorHandler1, customErrorHandler2);

        expect(window.onerror).toBe(customErrorHandler1);
        expect(window.unhandledrejection).toBe(customErrorHandler2);
    });
});
