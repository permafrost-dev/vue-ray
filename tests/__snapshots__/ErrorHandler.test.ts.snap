// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Error Handler: > sends an html payload and appends the callback result when handling vue errors 1`] = `
[
  {
    "content": "<pre><div class="text-black w-full block mt-1">Additional Info:</div><div class="text-blue-600 w-full block pl-8">---CUSTOM_DATA---</div></pre>",
    "label": "Error",
    "name": "custom",
  },
  {
    "content": "small",
    "label": null,
    "name": "small",
  },
  {
    "content": "red",
    "label": null,
    "name": "color",
  },
]
`;

exports[`Error Handler: > sends an html payload and appends the callback result when handling window errors 1`] = `
[
  {
    "content": "<pre>
        <div class="flex bg-white border-0">
            <div class="flex-row bg-white text-red-400 pr-1 pl-1 border border-1 border-red-400">message</div>
            <div style="padding-top: 0.25em;" class="flex-row"><code>source</code></div>
        </div><div class="text-black w-full block mt-1">Additional Info:</div><div class="text-blue-600 w-full block pl-8">---CUSTOM_DATA---</div></pre>",
    "label": "Error",
    "name": "custom",
  },
  {
    "content": "small",
    "label": null,
    "name": "small",
  },
  {
    "content": "red",
    "label": null,
    "name": "color",
  },
]
`;

exports[`Error Handler: > sends an html payload and appends the callback result when handling window unhandled rejection event errors 1`] = `
[
  {
    "content": "<pre>
        <div class="flex bg-white border-0">
            <div class="flex-row bg-white text-red-400 pr-1 pl-1 border border-1 border-red-400">some-data</div>
            <div style="padding-top: 0.25em;" class="flex-row"><code>{"reason":"some-data"}</code></div>
        </div><div class="text-black w-full block mt-1">Additional Info:</div><div class="text-blue-600 w-full block pl-8">---CUSTOM_DATA---</div></pre>",
    "label": "Error",
    "name": "custom",
  },
  {
    "content": "small",
    "label": null,
    "name": "small",
  },
  {
    "content": "red",
    "label": null,
    "name": "color",
  },
]
`;
